include: '../config.py'

rule all:
    input:
        ###
        # human
        ###
        TRANSCRIPTOME_FA,
        KALLISTO_INDEX,
        KALLISTO_SPIKEIN_INDEX,
        SALMON_INDEX,

        ###
        # mouse
        ###
        MOUSE_TRANSCRIPTOME_FA,
        MOUSE_KALLISTO_INDEX,
        MOUSE_SALMON_INDEX,

        ###
        # yeast
        ###
        YEAST_ANNOS,
        YEAST_TRANSCRIPTOME_FA,
        YEAST_KALLISTO_INDEX,
        YEAST_SALMON_INDEX,

        ###
        # spike-ins
        ###
        ERCC_FILE,
        ERCC_FA

rule get_transcriptome:
    output:
        TRANSCRIPTOME_FA
    shell:
        'curl -o gencode.v25.transcripts.fa.gz'
        ' --silent'
        ' ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_25/gencode.v25.transcripts.fa.gz && '
        ' gunzip gencode.v25.transcripts.fa.gz && '
        ' awk -F\'|\' \'{{print $1}}\' gencode.v25.transcripts.fa > {output} && '
        ' rm gencode.v25.transcripts.fa'

rule get_spikein_transcriptome:
    input:
        ERCC_FA,
        TRANSCRIPTOME_FA
    output:
        SPIKEIN_FA
    shell:
        ' '.join(['cat', TRANSCRIPTOME_FA, ERCC_FA, '>', SPIKEIN_FA])

rule kallisto_index:
    input:
        TRANSCRIPTOME_FA
    output:
        KALLISTO_INDEX
    shell:
        '{UPDATED_PATH} '
        'kallisto index'
        ' -i {output}'
        ' {input}'

rule salmon_index:
    input:
        TRANSCRIPTOME_FA
    threads:
        N_THREADS
    output:
        directory(SALMON_INDEX)
    shell:
        '{UPDATED_PATH} '
        'salmon index'
        ' -i {output}'
        ' -p {threads}'
        ' --keepDuplicates'
        ' -t {input}'

rule kal_spikein_index:
    input:
        SPIKEIN_FA
    output:
        KALLISTO_SPIKEIN_INDEX
    shell:
        '{UPDATED_PATH} '
        'kallisto index'
        ' -i {output}'
        ' {input}'

rule sal_spikein_index:
    input:
        SPIKEIN_FA
    threads:
        N_THREADS
    output:
        directory(SALMON_SPIKEIN_INDEX)
    shell:
        '{UPDATED_PATH} '
        'salmon index'
        ' -i {output}'
        ' -p {threads}'
        ' --keepDuplicates'
        ' -t {input}'

###
# mouse stuff
###
rule get_mouse_transcriptome:
    output:
        MOUSE_TRANSCRIPTOME_FA
    shell:
        'curl -o gencode.vM12.transcripts.fa.gz'
        ' --silent'
        ' ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M12/gencode.vM12.transcripts.fa.gz && '
        ' gunzip gencode.vM12.transcripts.fa.gz && '
        ' awk -F\'|\' \'{{print $1}}\' "gencode.vM12.transcripts.fa" > {output} && '
        ' rm gencode.vM12.transcripts.fa'

rule kallisto_mouse_index:
    input:
        MOUSE_TRANSCRIPTOME_FA
    output:
        MOUSE_KALLISTO_INDEX
    shell:
        '{UPDATED_PATH} '
        'kallisto index'
        ' -i {output}'
        ' -k 21'
        ' {input}'

rule salmon_mouse_index:
    input:
        MOUSE_TRANSCRIPTOME_FA
    threads:
        N_THREADS
    output:
        directory(MOUSE_SALMON_INDEX)
    shell:
        '{UPDATED_PATH} '
        'salmon index'
        ' -i {output}'
        ' -p {threads}'
        ' -k 21'
        ' --keepDuplicates'
        ' -t {input}'

###
# yeast stuff
###

rule get_yeast_transcriptome:
    output:
        YEAST_TRANSCRIPTOME_FA
    shell:
        'curl -o ASM294v2.pombase.cdna.fa.gz'
        ' --silent'
        ' ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/fasta/schizosaccharomyces_pombe/cdna/Schizosaccharomyces_pombe.ASM294v2.cdna.all.fa.gz && '
        ' gunzip ASM294v2.pombase.cdna.fa.gz && '
        'curl -o ASM294v2.pombase.ncrna.fa.gz'
        ' --silent'
        ' ftp://ftp.ensemblgenomes.org/pub/release-37/fungi/fasta/schizosaccharomyces_pombe/ncrna/Schizosaccharomyces_pombe.ASM294v2.ncrna.fa.gz && '
        ' gunzip ASM294v2.pombase.ncrna.fa.gz && '
        ' cat ASM294v2.pombase.cdna.fa ASM294v2.pombase.ncrna.fa > ASM294v2.pombase.all.fa && '
        ' rm ASM294v2.pombase.cdna.fa ASM294v2.pombase.ncrna.fa'

rule yeast_kallisto_index:
    input:
        YEAST_TRANSCRIPTOME_FA
    output:
        YEAST_KALLISTO_INDEX
    shell:
        '{UPDATED_PATH} '
        'kallisto index'
        ' -i {output}'
        ' {input}'

rule yeast_salmon_index:
    input:
        YEAST_TRANSCRIPTOME_FA
    threads:
        N_THREADS
    output:
        directory(YEAST_SALMON_INDEX)
    shell:
        '{UPDATED_PATH} '
        'salmon index'
        ' -i {output}'
        ' -p {threads}'
        ' --keepDuplicates'
        ' -t {input}'

rule yeast_annos:
    input:
        YEAST_TRANSCRIPTOME_FA
    output:
        YEAST_ANNOS
    shell:
        source_r('R', 'make_annos.R')

rule get_ercc:
    output:
        ERCC_FILE,
        ERCC_FA
    shell:
        'curl -o ERCC92_data.txt --silent'
        ' https://assets.thermofisher.com/TFS-Assets/LSG/manuals/cms_095046.txt'
        ' && '
        'curl -o ERCC92.zip --silent'
        ' https://assets.thermofisher.com/TFS-Assets/LSG/manuals/ERCC92.zip'
        ' && unzip ERCC92.zip && rm ERCC92.zip && '
        ' sed -i \'s/ERCC-/ERCC_/\' {output[1]}'
